### PR TITLE
Color graphs by utilizing CSP paradigmates

### PR DESCRIPTION
**Pull request by songofroland**

### CODE REVIEW

File: tests/test_graph_coloring.py [hunk 1]
```diff
@@ -5,64 +5,72 @@
 from .sample_graphs import SHACK_F_4, DI_3, STAR_S_6
 
 
+def assert_colored_elegantly(edges: list, vertices: list) -> None:
+    edges_colors = [el for row in edges for el in row if el != 0]
```
Comment (wilmhit) [inline tests/test_graph_coloring.py:9]: This line is unreadable

Comment (wilmhit) [reply tests/test_graph_coloring.py:9]: You can checkout graph_utils.py  from fix/validation if you need to count edges

File: tests/test_graph_coloring.py [hunk 2]
```diff
@@ -5,64 +5,72 @@
...
+    total_edges_colors_assigned = len(edges_colors) // 2
+    vertices_set = set(vertices)
+    edges_set = set(edges_colors)
```
Comment (wilmhit) [inline tests/test_graph_coloring.py:12]: I don't see a reason for using sets here. All operations below this point can be performed on lists as well.

Comment (wilmhit) [reply tests/test_graph_coloring.py:12]: Oh i see now

File: tests/test_graph_coloring.py [hunk 3]
```diff
@@ -5,64 +5,72 @@
 from .sample_graphs import SHACK_F_4, DI_3, STAR_S_6
 
 
+def assert_colored_elegantly(edges: list, vertices: list) -> None:
```
Comment (wilmhit) [inline tests/test_graph_coloring.py:8]: Use AdjMatrix type for edges

**Review by wilmhit — CHANGES_REQUESTED** [review]

File: tests/test_graph_coloring.py (hunk 1 above)
Comment (wilmhit) [reply tests/test_graph_coloring.py:9]: you could use `filter` function here

File: graph_coloring/graph_utils.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+from .custom_types import AdjMatrix
+
+
+def count_edges(graph: AdjMatrix) -> int:
+    return sum(sum(element > 0 for element in row) for row in graph) // 2
```
Comment (songofroland) [inline graph_coloring/graph_utils.py:5]: Your method was really interesting but a little bit overcomplicated. As a daily Python user, I'd say this one is more readable.

File: tests/test_graph_coloring.py (hunk 1 above)
Comment (songofroland) [reply tests/test_graph_coloring.py:9]: thanks for the idea, although in my opinion list comprehension solution is more readable
when you'll get more experience with Python, you'll feel very comfortable with using them

image

Comment (songofroland) [reply tests/test_graph_coloring.py]: I'm not stubborn tho, if u insist to use the second one then I could adjust it

Comment (wilmhit) [reply tests/test_graph_coloring.py]: List comprehension are fine. It's just this one is very complicated.

Comment (wilmhit) [reply tests/test_graph_coloring.py]: How about:
```python
filter(lambda color: color !=0,  [color for color in row for row in edges])
```
but ultimately it's your choice

**Review by wilmhit — APPROVED** [review]

**Pull request merged**